### ISSUE TITLE
serialisation from one class type to another class if both have exactly same properties

### ISSUE THREAD

**Issue by neerajyadav** [google/built_value.dart#541]
Hi, I'm developing a `flutter` project with BLoC architecture. To re-use the business logic I split my project in three components.
- `bloc` :  pure `dart library`, this has all the business logic including `models`, `entities`, `repository` and `BLoC`. `repository` here is an abstract definition whose implementation is in respective `app` and `web` component.
- `app` : `flutter` project, which has app specific implementation of `repository`
- `web` : `angularDart` project which has web code and implementation of `repository`.

I had to split `repository` across `bloc` and app/web because I'm using `firestore` as backend db and `firestore` has different `pub package` for web and `flutter`.

My documents has [`GeoPoint`](https://firebase.google.com/docs/reference/android/com/google/firebase/firestore/GeoPoint) and [`Timestamp`](https://firebase.google.com/docs/reference/js/firebase.firestore.Timestamp) fields.
I can't access `GeoPoint` on my `bloc` dart library as it does not have `cloud_firestore` pub package. So I created exact same class in `bloc` dart lib with same name and properties and use it as type in my `entities`. Now when I'm serialising my `entity` it is giving me error.
`UserEntity`
``` dart
abstract class UserEntity
    with BaseEntity
    implements Built<UserEntity, UserEntityBuilder> {
  UserEntity._();

  factory UserEntity([updates(UserEntityBuilder b)]) = _$UserEntity;

  @nullable
  @BuiltValueField(wireName: 'name')
  String get name;

  @nullable
  @BuiltValueField(wireName: 'username')
  String get username;

  @nullable
  @BuiltValueField(wireName: 'email')
  String get email;

  @nullable
  @BuiltValueField(wireName: 'emailVerified')
  bool get emailVerified;

  @nullable
  @BuiltValueField(wireName: 'phone')
  String get phone;

  @nullable
  @BuiltValueField(wireName: 'photo')
  String get photo;

  @nullable
  @BuiltValueField(wireName: 'thumbnail')
  String get thumbnail;

  @nullable
  @BuiltValueField(wireName: 'location')
  GeoPoint get location;

  String toJson() {
    return json.encode(serializers.serializeWith(UserEntity.serializer, this));
  }

  static UserEntity fromJson(String jsonString) {
    return serializers.deserializeWith(
        UserEntity.serializer, json.decode(jsonString));
  }

  static Serializer<UserEntity> get serializer => _$userEntitySerializer;
}

``` 
When I serialise with this code

``` dart
var user = standardSerializers.deserializeWith<UserEntity>(UserEntity.serializer, snapshot.data);
``` 

``` dart
final Serializers serializers = _$serializers;
final standardSerializers =
    (serializers.toBuilder()
      ..add(GeoPointSerializer())
      ..addPlugin(new StandardJsonPlugin())).build();
``` 
I'm getting this error:
``` log
Deserializing '[name, Neeraj, username, neeraj, phone, xxxxx, photo, https://lh5....' to 'UserEntity' failed due to: Deserializing '[location, Instance of 'GeoPoint', null, null, null, null, null, null]' to 'Place' failed due to: Deserializing 'Instance of 'GeoPoint'' to 'GeoPoint' failed due to: type 'GeoPoint' is not a subtype of type 'GeoPoint' in type cast where
GeoPoint is from package:cloud_firestore/cloud_firestore.dart
GeoPoint is from package:bloc/src/models/geo_point.dart
#0      BuiltJsonSerializers._deserialize (package:built_value/src/built_json_serializers.dart:151:11)
#1      BuiltJsonSerializers.deserialize (package:built_value/src/built_json_serializers.dart:102:18)
#2      BuiltJsonSerializers.deserializeWith (package:built_value/src/built_json_serializers.dart:32:12)
#3      UserRepositoryImpl.getUser.<anonymous closure> (package:travel_buddy_app/src/data/user_repository_impl.dart:24:38)
#4     <…>
```

**Comment by davidmorgan** [comment]
It looks like you have a custom `GeoPoint` serialize that just passes the value through?

One way to do this would be to make it more like the generated `GeoPoint` serializer for the class you wrote by hand. It needs to pull out the fields from `GeoPoint` and return them serialized.

Then, you'll need two different `serializer` instances: one which uses the real `GeoPoint` and one which uses your hand written `GeoPoint`. They should then be compatible, because the fields are the same.

By the way, you don't need to write `@BuiltValueField(wireName: ...)` if the wire name matches the field name; that's the default.

**Comment by neerajyadav** [comment]
Thanks @davidmorgan for the quick reply. 
This was my previous `GeoPointSerializer`

``` dart
class GeoPointSerializer implements PrimitiveSerializer<GeoPoint> {
  final bool structured = false;
  @override
  final Iterable<Type> types = new BuiltList<Type>([GeoPoint]);
  @override
  final String wireName = 'GeoPoint';

  @override
  Object serialize(Serializers serializers, GeoPoint geoPoint,
      {FullType specifiedType: FullType.unspecified}) {
    return geoPoint;
  }

  @override
  GeoPoint deserialize(Serializers serializers, Object serialized,
      {FullType specifiedType: FullType.unspecified}) {
    return serialized as GeoPoint;
  }
}
```
This one is new,
``` dart
class GeoPointSerializer implements StructuredSerializer<GeoPoint> {
  @override
  GeoPoint deserialize(Serializers serializers, Iterable serialized,
      {FullType specifiedType = FullType.unspecified}) {
    final iterator = serialized.iterator;
    double latitude;
    double longitude;
    while (iterator.moveNext()) {
      final key = iterator.current as String;
      iterator.moveNext();
      final dynamic value = iterator.current;
      switch (key) {
        case 'latitude':
          latitude = serializers.deserialize(value,
              specifiedType: const FullType(double)) as double;
          break;
        case 'total_trips':
          longitude = serializers.deserialize(value,
              specifiedType: const FullType(double)) as double;
          break;
      }
    }
    return GeoPoint(latitude, longitude);
  }

  @override
  Iterable serialize(Serializers serializers, GeoPoint object,
      {FullType specifiedType = FullType.unspecified}) {
    final result = <Object>[];
    if (object.latitude != null) {
      result
        ..add('latitude')
        ..add(serializers.serialize(object.latitude,
            specifiedType: const FullType(double)));
    }
    if (object.longitude != null) {
      result
        ..add('longitude')
        ..add(serializers.serialize(object.latitude,
            specifiedType: const FullType(double)));
    }
    return result;
  }

  @override
  Iterable<Type> get types => const [GeoPoint];

  @override
  String get wireName => "GeoPoint";
}
```

I added `GeoPointSerializer` in list the same earlier.
``` dart
@SerializersFor(const [AuthUser, UserEntity, GeoPoint])
final Serializers serializers = _$serializers;
final standardSerializers =
    (serializers.toBuilder()
      ..add(DateTimeSerializer())
      ..add(GeoPointSerializer())
      ..addPlugin(new StandardJsonPlugin())).build();
``` 

Now I'm getting this error.
``` log
'UserEntity' failed due to: Deserializing '[location, Instance of 'GeoPoint', null, null, null, null, null, null]' to 'Place' failed due to: Deserializing 'Instance of 'GeoPoint'' to 'GeoPoint' failed due to: type 'GeoPoint' is not a subtype of type 'Iterable<dynamic>' in type cast
#0      BuiltJsonSerializers._deserialize (package:built_value/src/built_json_serializers.dart:151:11)
#1      BuiltJsonSerializers.deserialize (package:built_value/src/built_json_serializers.dart:102:18)
#2      BuiltJsonSerializers.deserializeWith (package:built_value/src/built_json_serializers.dart:32:12)
#3      UserEntity.fromDocument (package:travel_buddy_bloc/src/models/user_entity.dart:93:32)
#4      UserRepositoryImpl.getUser.<anonymous closure> (package:travel_buddy_app/src/data/user_repository_impl.dart:24:29)
#5      _RootZone.runUnary (dart:async/zone.da<…>
```

**Comment by davidmorgan** [comment]
It looks like whatever is doing the serialization is still using the old serializer.

**Comment by neerajyadav** [comment]
@davidmorgan I completely removed the old serializer from code. Earlier I was getting error
``` error
type 'GeoPoint' is not a subtype of type 'GeoPoint' in type cast
```

And now I'm getting
``` error
type 'GeoPoint' is not a subtype of type 'Iterable<dynamic>'
```

**Comment by davidmorgan** [comment]
Sorry, this is more complicated than I said.

What I think you need to do is in the code that does have access to GeoPoint, process the data to convert the real `GeoPoint` into your version. You could do that by recursively processing all the maps. Then, the serializer should work with the generated `GeoPoint` serializer.

**Comment by neerajyadav** [comment]
Thanks @davidmorgan, my app is around travel domain and majority of operations include GeoPoint, every time I fetch data from `firestore` I need to do the conversion. I think that will prove quite costly in terms of performance. Is this the only way to solve it?

**Comment by davidmorgan** [comment]
It might be that you could do the conversion in a serializer plugin--something like StandardJsonPlugin--that would save you going through all the data yourself.

**Comment by neerajyadav** [comment]
I tried to write a plugin like `StandardJsonPlugin`. The important part is, in `beforeDeserialize` method what I tried is
``` dart
if (specifiedType.root.toString() == "GeoPoint") {
      final geopoint = json.encode(object);
      GeoPoint point  = json.decode(geopoint);
      return point;
    }
```
However looks like `Firebase GeoPoint` is not encodable at all. This is the error I'm getting.
``` log
Converting object to an encodable object failed: Instance of 'GeoPoint'
```
Is there a way I can fetch `latitude` and `longitude` properties from `Object object`? If I try to cast it as my custom `GeoPoint` class, it gives me same casting error.

I need help in fetching `latitude` and `longitude` properties from `Object object`.
